src/sources/closed_coil.cpp

```diff
@@ -0,0 +1,634 @@
+    sps_params_[i]->SetParam("SourceName",
+                             std::string("source_" + std::to_string(i)));
+    sps_params_[i]->SetParam("PotentialName",
+                             std::string("V_" + std::to_string(i)));
+    sps_params_[i]->SetParam("HCurlFESpaceName",
+                             std::string("HCurl_" + std::to_string(i)));
+    sps_params_[i]->SetParam("H1FESpaceName",
+                             std::string("H1_" + std::to_string(i)));
+    sps_params_[i]->SetParam("SolverOptions", *current_solver_options_[i]);
+    sps_params_[i]->SetParam("ConductivityCoefName",
+                             std::string("magnetic_permeability"));
+
+    coefs_[i] = new hephaestus::Coefficients;
+    coefs_[i]->scalars.Register("magnetic_permeability", coef1_, false);
+
+    sps_[i] = new hephaestus::ScalarPotentialSource(*sps_params_[i]);
+    sps_[i]->Init(*gridfunctions_[i], *fespaces_[i], *bc_maps_[i], *coefs_[i]);
+
+    mfem::ParLinearForm dummy(HCurlFESpace_[i]);
+    sps_[i]->Apply(&dummy);
+
+    if (i == 1){
+      src_J_ = new mfem::VectorGridFunctionCoefficient(J_[0]);
+      high_DBC_J_ = new hephaestus::VectorFunctionDirichletBC(
+          std::string("source_" + std::to_string(i)), high_terminal_, src_J_);
+      low_DBC_J_ = new hephaestus::VectorFunctionDirichletBC(
+          std::string("source_" + std::to_string(i)), low_terminal_, src_J_);
+      bc_maps_[i]->Register("high_current", high_DBC_J_, true);
+      bc_maps_[i]->Register("low_current", low_DBC_J_, true);
+    }
+
+    // Clean the divergence of the two J fields
+    cleanDivergence(gridfunctions_[i], std::string("source_" + std::to_string(i)), std::string("V_" + std::to_string(i)), bc_maps_[i]);
+
+    // Normalise the current through the wedges and use them as a reference
+    // The MPI allreduce is to take into account the MPI partitioning
+    double flux = calcFlux(J_[i], elec_attrs_.first);
+    double total_flux;
+    MPI_Allreduce(&flux, &total_flux, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+    *J_[i] /= abs(total_flux);
+  }
+
+  *J_[0] *= -1.0;
+}
+
+void ClosedCoilSolver::cleanDivergence(hephaestus::GridFunctions* gridfunctions, std::string J_name, std::string V_name, hephaestus::BCMap* bc_map){
+
+  hephaestus::InputParameters pars;
+  hephaestus::FESpaces fes;
+
+  pars.SetParam("VectorGridFunctionName",
+                             J_name);
+  pars.SetParam("ScalarGridFunctionName",
+                             V_name);
+  hephaestus::HelmholtzProjector projector(pars);
+  projector.Project(*gridfunctions, fes, *bc_map);
+
+}
+
+void ClosedCoilSolver::restoreAttributes() {
+
+  // Domain attributes
+  for (int e = 0; e < mesh_parent_->GetNE(); ++e) {
+    mesh_parent_->SetAttribute(e, old_dom_attrs[e]);
+  }
+
+  mesh_parent_->FinalizeTopology();
+  mesh_parent_->Finalize();
+  mesh_parent_->SetAttributes();
+}
+
+// Auxiliary methods
+
+double calcFlux(mfem::GridFunction *v_field, int face_attr) {
```

Comment:
Worth documenting (or moving here) the `MPI_Allreduce` call? Looks like it'd be easy for a user to miss it and only sum the flux over the local rank?

Reply:
That's a very good point, yes. I'll do this!